[ML] Do not create inference endpoint if ID is used in existing mappings

DonalEvans opened the pull request:
When creating an inference endpoint, if the inference ID is referenced by ingest pipeline processors or used in semantic_text mappings in non-empty indices, prevent the endpoint from being created.

Closes #124272

DonalEvans committed:
Do not create inference endpoint if ID is used in existing mappings

When creating an inference endpoint, if the inference ID is referenced
by ingest pipeline processors or used in semantic_text mappings in
non-empty indices, prevent the endpoint from being created.

Closes #124272

DonalEvans committed:
Update docs/changelog/137055.yaml

elasticsearchmachine commented:
Hi @donalevans, I've created a changelog YAML for you.

DonalEvans committed:
Reword error message

DonalEvans committed:
Merge branch 'main' into check-existing-semantic-text-fields-when-creating-inference-endpoint

DonalEvans committed:
Fix failing test

DonalEvans commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportPutInferenceModelAction.java:
```diff
@@ -221,7 +246,72 @@ private void parseAndStoreModel(
+                    nonEmptyIndices.add(indexName);
+                }
+                searchResponse.decRef();
+            }
```
Thanks for the explanation! So it sounds like we're okay with the possibility of a user creating an endpoint that isn't compatible with existing mappings (if the model is different, for example), but that we should at least check those model settings that we do have access to.

I'll implement that logic for indices with semantic text fields, but I'm not sure if it's possible to do the same kind of fine-grained check for pipelines, since as far as I can tell, they only store the inference endpoint ID and not any of the model settings. Would preventing endpoint creation in the event that any pipelines exist which reference the endpoint ID be acceptable, or is that too heavy-handed, and we should dispense with checking the pipelines altogether?

benwtrent replied:
> Would preventing endpoint creation in the event that any pipelines exist which reference the endpoint ID be acceptable, or is that too heavy-handed, and we should dispense with checking the pipelines altogether?

I think we dispense with checking pipelines altogether. Pipelines and indices are completely separate from each other. Consequently, one shouldn't impact the other I don't think. At least, not in the case.

DonalEvans committed:
Check model settings for existing semantic text fields

- Do not check for pipelines using the inference ID
- Check if existing semantic text fields have compatible model settings
- Update and expand test coverage for the new behaviour
- Improve existing test InferenceServiceExtension implementations
- Move SemanticTextInfoExtractor from xpack.core.ml.utils to
  xpack.inference.common

DonalEvans committed:
Merge branch 'main' into check-existing-semantic-text-fields-when-creating-inference-endpoint

benwtrent commented on x-pack/plugin/inference/qa/inference-service-tests/src/javaRestTest/java/org/elasticsearch/xpack/inference/InferenceCrudIT.java:
```diff
@@ -303,6 +304,74 @@ public void testDeleteEndpointWhileReferencedBySemanticTextAndPipeline() throws
+        ResponseException responseException = assertThrows(
+            ResponseException.class,
+            () -> putModel(endpointId, mockDenseServiceModelConfig(128), TaskType.TEXT_EMBEDDING)
+        );
```
Seems like this should fail even if there aren't any docs in the index.

DonalEvans commented on server/src/main/java/org/elasticsearch/inference/MinimalServiceSettings.java:
```diff
@@ -62,9 +62,9 @@ public record MinimalServiceSettings(
-    static final String ELEMENT_TYPE_FIELD = "element_type";
+    public static final String DIMENSIONS_FIELD = "dimensions";
+    public static final String SIMILARITY_FIELD = "similarity";
+    public static final String ELEMENT_TYPE_FIELD = "element_type";
```
My mistake, I was originally using these in tests, but then realized that `ServiceFields` has the same Strings defined and already public. I'll put them back how they were. It would be nice if we could have a single source of truth for these sorts of constants, because they're defined independently in a lot of different places, which makes it a real pain if we ever want to update them.

DonalEvans committed:
Restore default visibility to MinimalServiceSettings constants

DonalEvans committed:
Merge branch 'main' into check-existing-semantic-text-fields-when-creating-inference-endpoint
